### PR TITLE
BFD+VXLAN Automation

### PR DESCRIPTION
**Pull request by gollarharsha**
BFD automation with VXLAN was done for the first 3 cases:
 
Here are additional changes that were added:
 
As discussed in last session, I have enabled BFD in the same file as VXLAN. By passing bfd argument, we should be able to run existing VXLAN test cases.
Bfd_sniffer script will extract BFD packets and automatically populate the config file for bfd_responder.
Cleanup for BFD was added in existing vxlan script itself.
 
I have ran first 3 test cases for all 4 encap types. All 12 cases passed as well.

### CODE REVIEW

**Comment by gollarharsha** [comment]
FYI - @siqbal1986

File: tests/vxlan/test_vxlan_ecmp.py [hunk 1]
```diff
@@ -700,6 +248,7 @@ def test_vxlan_single_endpoint(self, setUp, encap_type):
             tc1:Create a tunnel route to a single endpoint a. Send packets to the route prefix dst.
         '''
         self.setup = setUp
+        import pdb; pdb.set_trace();
```
Comment (siqbal1986) [inline tests/vxlan/test_vxlan_ecmp.py:251]: this should be removed.

Comment (gollarharsha) [reply tests/vxlan/test_vxlan_ecmp.py]: Removed it and pushed changes along with additional test case

**Comment by siqbal1986** [comment]
Can you please specify which Vxlan tests you ran with the BFD exactly?

**Comment by gollarharsha** [comment]
Hi @siqbal1986 , 

**Here are the Test cases that will work with BFD:**
**Test_VxLAN_route_tests**
test_vxlan_single_endpoint
test_vxlan_modify_route_different_endpoint
test_vxlan_remove_all_route

**Test_VxLAN_ecmp_create**
test_vxlan_configure_route1_ecmp_group_a
test_vxlan_configure_route1_ecmp_group_b
test_vxlan_configure_route2_ecmp_group_b

**Review by ihorchekh — CHANGES_REQUESTED** [review]
Please review these changes

File: tests/vxlan/vxlan_ecmp_utils.py [hunk 2]
```diff
@@ -0,0 +1,599 @@
...
+        #Finding ethernet interfaces under that specific portchannel
+        required_ethernet_interfaces = minigraph_facts['minigraph_portchannels'][required_pc]['members']
+        #Finding interfaces with PTF indices
+        ret_list = []
+        for iface in required_ethernet_interfaces:
+            ret_list.append(minigraph_facts["minigraph_ptf_indices"][iface])
+        return ret_list
+
+    def ptf_config(self, duthost, ptfhost, tbinfo, delete_member_a1 = None, delete_member_a2 = None):
+        responder_output = ptfhost.command('supervisorctl stop bfd_responder')
```
Comment (ihorchekh) [inline tests/vxlan/vxlan_ecmp_utils.py:577]: When bfd_responder is not running on PTF, we're getting an error here.
`bfd_responder: ERROR (no such process)`
Solution:
```suggestion
        responder_output = ptfhost.command('supervisorctl stop bfd_responder', module_ignore_errors=True)
```

**Review by ihorchekh — CHANGES_REQUESTED** [review]

File: tests/vxlan/bfd_sniffer.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
+from scapy.all import *
```
Comment (ihorchekh) [inline tests/vxlan/bfd_sniffer.py:1]: Please reformat this file according to PEP8.

**Comment by ihorchekh** [comment]
This PR has too many issues. I tried to add some small fixes to make it run, but I couldn't. Please fix those comments I added and make sure the test is working stable.

**Comment by siqbal1986** [comment]
@gollarharsha Can you please address the issues raised by @ihorchekh

File: tests/vxlan/conftest.py [hunk 4]
```diff
@@ -1,15 +1,25 @@
-
+import argparse 
+import pytest
 import logging
-import yaml
-
+logger = logging.getLogger(__name__)
```
[the remark this answers is not in the document]
Comment (gollarharsha) [reply tests/vxlan/conftest.py:4]: Removed it

File: tests/vxlan/conftest.py [hunk 5]
```diff
@@ -134,99 +187,3 @@ def pytest_addoption(parser):
         type=int,
         help="ECMP: Number of tunnel nexthops to be tested. (number of nhs_per_destination X number_of_destinations)"
     )
-
```
[the remark this answers is not in the document]
Comment (gollarharsha) [reply tests/vxlan/conftest.py:137]: I have added them back

File: tests/vxlan/test_vxlan_ecmp.py [hunk 6]
```diff
@@ -537,37 +69,44 @@ def encap_type(request):
 @pytest.fixture(scope="module")
 def setUp(duthosts, ptfhost, request, rand_one_dut_hostname, minigraph_facts,
           tbinfo, encap_type):
+    
+    asic_type = duthosts[rand_one_dut_hostname].facts["asic_type"]
+    if asic_type == "cisco-8000":
+        ECMP_TOLERANCE = 0.01
+    else:
+        ECMP_TOLERANCE = 0.01
```
[the remark this answers is not in the document]
Comment (gollarharsha) [reply tests/vxlan/test_vxlan_ecmp.py]: Removed the else condition for now. But we other platform may need to modify this based on their tolerance

**Comment by kevinskwang** [comment]
@gollarharsha please check the pre-commit-check failures and fix it.

**Comment by rraghav-cisco** [comment]
Adding logs from the last changes.
[script_logs.tar.gz](https://github.com/sonic-net/sonic-mgmt/files/9883374/script_logs.tar.gz)
@siqbal1986 , pls let us know how it goes.

**Comment by rraghav-cisco** [comment]
@siqbal1986 
1. This new update includes fixing the pre-commit errors.
2. Fixing the time to run: By default the script will run the quick tests, which take a total of 45 minutes for 4 encap types.
3. If "--include_long_tests=True" is provided, the total run time is 3:15 hours.
4. CRM is still excluded.
5. Updated the ptf traffic script to send all packets quickly and then wait a fixed time for all the return packets. This greatly reduced the time to run for 1000 packets from more than 10 minutes to less than 2 minutes.

**Comment by rraghav-cisco** [comment]
@siqbal1986, pls merge it.

**Comment by prsunny** [comment]
@siqbal1986 , is this planned to be cherry-picked to 202012 or someone raising a separate PR?

**Comment by siqbal1986** [comment]
> @siqbal1986 , is this planned to be cherry-picked to 202012 or raising a separate PR?

I don't think automatic cherry pick on this would work. @gollarharsha @rraghav-cisco I think a separate PR would be required. Can you please raise one

**Comment by rraghav-cisco** [comment]
@siqbal1986 and @prsunny, I will take it up.

**Comment by prsunny** [comment]
@ihorchekh , can you please check if all comments were addressed and signoff?

File: tests/vxlan/bfd_sniffer.py (hunk 3 above)
Comment (rraghav-cisco) [reply tests/vxlan/bfd_sniffer.py:5]: @ihorchekh , we are planning to remove this file, as it is not needed for the tests.

**Comment by rraghav-cisco** [comment]
@prsunny , @siqbal1986 : I have raised https://github.com/sonic-net/sonic-mgmt/pull/6807

**Pull request merged**